File: docs/nifi-analyzer/src/parsers/nel/tokenizer.js
// ================================================================
// nel/tokenizer.js — Tokenize NEL chained expressions
// Extracted from monolith lines 3187-3210
// ================================================================

/**
 * Split a NiFi Expression Language chain on top-level colons,
 * respecting nested parentheses and quoted strings.
 *
 * Example: "filename:substringBefore('.'):toUpper()" -> ["filename", "substringBefore('.')", "toUpper()"]
 *
 * @param {string} expr - Raw NEL expression (inside ${...})
 * @returns {string[]} Array of chain segments
 */
export function tokenizeNELChain(expr) {
  var parts = []; var current = ''; var depth = 0; var inStr = false; var strChar = '';
  for (var i = 0; i < expr.length; i++) {
    var ch = expr[i];
    if (inStr) {
      current += ch;
      if (ch === strChar) inStr = false;
    } else if (ch === "'" || ch === '"') {
      inStr = true; strChar = ch; current += ch;
    } else if (ch === '(') {
      depth++; current += ch;
    } else if (ch === ')') {
      depth--; current += ch;
    } else if (ch === ':' && depth === 0) {
      // Always split on top-level colons — each segment is a chain step.
      // Namespace prefixes (math:floor, etc.) are handled at the parser level.
      parts.push(current); current = '';
    } else {
      current += ch;
    }
  }
  if (current) parts.push(current);
  return parts;
}

File: docs/nifi-analyzer/backend/app/engines/parsers/nel/tokenizer.py
"""NEL tokenizer — split NiFi Expression Language chains on top-level colons.

Ported from nel/tokenizer.js.
"""


def tokenize_nel_chain(expr: str) -> list[str]:
    """Split a NEL chain on top-level colons, respecting nested parens and quotes.

    Example: "filename:substringBefore('.'):toUpper()" -> ["filename", "substringBefore('.')", "toUpper()"]
    """
    parts: list[str] = []
    current = ""
    depth = 0
    in_str = False
    str_char = ""
    escaped = False

    for ch in expr:
        if escaped:
            # Previous char was backslash; consume this char literally
            current += ch
            escaped = False
        elif ch == "\\" and in_str:
            # Backslash inside a string — next char is escaped
            current += ch
            escaped = True
        elif in_str:
            current += ch
            if ch == str_char:
                in_str = False
        elif ch in ("'", '"'):
            in_str = True
            str_char = ch
            current += ch
        elif ch == "(":
            depth += 1
            current += ch
        elif ch == ")":
            depth -= 1
            current += ch
        elif ch == ":" and depth == 0:
            parts.append(current)
            current = ""
        else:
            current += ch

    if current:
        parts.append(current)
    return parts
